include/psa/crypto_extra.h

```diff
@@ -711,6 +711,99 @@ psa_status_t mbedtls_psa_external_get_random(
 
 /**@}*/
 
+/** \defgroup psa_builtin_keys Built-in keys
+ * @{
+ */
+
+/** The minimum value for a key identifier that is built into the
+ * implementation.
+ *
+ * The range of key identifiers from #MBEDTLS_PSA_KEY_ID_BUILTIN_MIN
+ * to #MBEDTLS_PSA_KEY_ID_BUILTIN_MAX within the range from
+ * #PSA_KEY_ID_VENDOR_MIN and #PSA_KEY_ID_VENDOR_MAX and must not intersect
+ * with any other set of implementation-chosen key identifiers.
+ *
+ * This value is part of the library's ABI since changing it would invalidate
+ * the values of built-in key identifiers in applications.
+ */
+#define MBEDTLS_PSA_KEY_ID_BUILTIN_MIN          ((psa_key_id_t)0x7fff0000)
+
+/** The maximum value for a key identifier that is built into the
+ * implementation.
+ *
+ * See #MBEDTLS_PSA_KEY_ID_BUILTIN_MIN for more information.
+ */
+#define MBEDTLS_PSA_KEY_ID_BUILTIN_MAX          ((psa_key_id_t)0x7fffefff)
+
+/** A slot number identifying a key in a driver.
+ *
+ * Values of this type are used to identify built-in keys.
+ */
+typedef uint64_t psa_drv_slot_number_t;
```

Comment:
Unsure. This is what @gilles-peskine-arm set up in #3822...

Reply:
The choice of type is meant to accommodate different ways of selecting slot numbers.

There are secure elements with a very small number of physical slots where dynamic allocation is not realistic: each slot is assigned a purpose at device integration time. This is usually the case on M-class devices with a secure element. On these devices, `uint8_t` would be plenty.

In larger applications, keys can be created in secure elements dynamically. Many secure elements do not have a way to choose a free “slot number” when creating a key. For example, if the secure element itself exposes a PSA crypto API interface, the ”slot number“ would naturally be the key id on the secure element, which the caller has to choose. With PKCS#11, the “slot number” would naturally be a label which the caller has to choose.

There are several potential ways to create a unique identifier when you don't know what identifiers have already been used. You can generate a random identifier, but it needs to be long enough in order for the probability of repeats to be negligible, much more than 64 bits. You can use a counter, which requires maintaining state, but guarantees uniqueness. A 32-bit counter can realistically wrap during the lifetime of a device, especially if a malicious actor (such as a breached application on a multi-application system) is allowed to create keys. A 64-bit counter will not wrap if your device's life span is at most a few decades.

Hence the choice of 64 bits, which is large enough to allow a simple way of choosing slot numbers while still being easy to manipulate (in particular, it's fixed-size).

> Having to handle uint64_t on embedded devices without CPU support does have a significant cost in terms of performance and code size.

How is this worse than `uint32_t[2]` or `uint8_t[8]`?

Reply:
If you use the uint64_t as an `uint32_t[2]` or `uint8_t[8]` that's not worse AFAIK. It is when you start to do 64 bits arithmetic on a 32 bits processor that there is a performance and code size cost compared to just 32 bits arithmetic. Not sure we expect much arithmetic to be done on slot numbers thus it could be all right.
Another possibility as the needs seem quite different depending on the platform would be for this type to be platform dependent. As long as the core does not have to do arithmetic and comparison with slot numbers this shouldn't be to painful to handle.

Reply:
Since there's currently no arithmetic being done on this type by the core (and comparison only happens inside of the test driver), are you okay with this PR using the proposed uint64_t for the reasons outlined by Gilles, @ronald-cron-arm? The type can always be turned into a platform specific type at a later point in time in case the little overhead turns out to be disruptive to a platform implementer.

Reply:
Yes, that's fine by me.